@alexmoore ITestClusterLifecycle.testManyCommandsOneConnection() fails on my laptop. Here is a log

```
2016-04-04 13:47:12 [nioEventLoopGroup-2-1] DEBUG com.basho.riak.client.core.RiakNode - Riak replied with error; 0:no_type
2016-04-04 13:47:12 [nioEventLoopGroup-2-1] DEBUG com.basho.riak.client.core.RiakCluster - operation failed; remaining retries: 2
2016-04-04 13:47:12 [pool-1-thread-1] DEBUG com.basho.riak.client.core.RiakNode - Attempting to acquire channel permit
2016-04-04 13:47:12 [pool-1-thread-1] DEBUG com.basho.riak.client.core.RiakNode - Operation not being executed Riaknode 127.0.0.1:10017; no connections available
2016-04-04 13:47:12 [pool-1-thread-1] DEBUG com.basho.riak.client.core.RiakCluster - operation failed; remaining retries: 1
2016-04-04 13:47:12 [pool-1-thread-1] DEBUG com.basho.riak.client.core.RiakNode - Attempting to acquire channel permit
2016-04-04 13:47:12 [nioEventLoopGroup-2-1] DEBUG com.basho.riak.client.core.RiakNode - Channel id:300790552 returned to pool
2016-04-04 13:47:12 [pool-1-thread-1] DEBUG com.basho.riak.client.core.RiakNode - Operation not being executed Riaknode 127.0.0.1:10017; no connections available
2016-04-04 13:47:12 [nioEventLoopGroup-2-1] DEBUG com.basho.riak.client.core.RiakNode - Released pool permit
2016-04-04 13:47:12 [pool-1-thread-1] DEBUG com.basho.riak.client.core.FutureOperation - Setting Complete on future
2016-04-04 13:47:12 [pool-1-thread-1] DEBUG com.basho.riak.client.core.RiakCluster - operation failed; remaining retries: 0
2016-04-04 13:47:31 [main] DEBUG com.basho.riak.client.api.ITestClusterLifecycle - Exception occurred
java.util.concurrent.ExecutionException: com.basho.riak.client.core.NoNodesAvailableException
    at com.basho.riak.client.core.FutureOperation.get(FutureOperation.java:314)
    at com.basho.riak.client.api.commands.CoreFutureAdapter.get(CoreFutureAdapter.java:52)
    at com.basho.riak.client.api.ITestClusterLifecycle.createAndStoreObject(ITestClusterLifecycle.java:86)
    at com.basho.riak.client.api.ITestClusterLifecycle.testManyCommandsOneConnection(ITestClusterLifecycle.java:64)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
    at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:119)
    at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
    at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
    at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
Caused by: com.basho.riak.client.core.NoNodesAvailableException
    at com.basho.riak.client.core.RiakCluster.retryOperation(RiakCluster.java:469)
    at com.basho.riak.client.core.RiakCluster.access$1000(RiakCluster.java:48)
    at com.basho.riak.client.core.RiakCluster$RetryTask.run(RiakCluster.java:554)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)
2016-04-04 13:47:31 [main] DEBUG com.basho.riak.client.api.ITestClusterLifecycle - Cluster state: RUNNING, iteration: 0
2016-04-04 13:47:31 [main] INFO com.basho.riak.client.core.RiakCluster - RiakCluster is shutting down.
```

--- a/pom.xml
+++ b/pom.xml
@@ -338,6 +338,12 @@
             <scope>test</scope>
         </dependency>
         <dependency>
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest-core</artifactId>
+            <version>1.3</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
             <groupId>com.jayway.awaitility</groupId>
             <artifactId>awaitility</artifactId>
             <version>1.3.5</version>

--- a/src/main/java/com/basho/riak/client/core/RiakNode.java
+++ b/src/main/java/com/basho/riak/client/core/RiakNode.java
@@ -889,8 +889,8 @@ public class RiakNode implements RiakResponseListener
         consecutiveFailedOperations.incrementAndGet();
         if (inProgress != null)
         {
-            inProgress.setException(ex);
             returnConnection(channel); // release permit
+            inProgress.setException(ex);
         }
     }
 

--- a/src/test/java/com/basho/riak/client/api/ITestClusterLifecycle.java
+++ b/src/test/java/com/basho/riak/client/api/ITestClusterLifecycle.java
@@ -7,7 +7,9 @@ import com.basho.riak.client.core.RiakFuture;
 import com.basho.riak.client.core.RiakNode;
 import com.basho.riak.client.core.query.Location;
 import com.basho.riak.client.core.query.Namespace;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.ExpectedException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -16,6 +18,7 @@ import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 
+import static org.hamcrest.core.StringContains.containsString;
 import static org.junit.Assert.*;
 import static org.junit.Assume.assumeTrue;
 
@@ -28,6 +31,8 @@ public class ITestClusterLifecycle
     protected static Random random = new Random();
     private final Logger logger = LoggerFactory.getLogger(ITestClusterLifecycle.class);
 
+    @Rule
+    public ExpectedException thrown= ExpectedException.none();
 
     public ITestClusterLifecycle()
     {
@@ -69,7 +74,32 @@ public class ITestClusterLifecycle
         {
             logger.debug("Exception occurred", ex);
             logger.debug("Cluster state: {}, iteration: {}", cluster.getNodes().get(0).getNodeState().toString(), i);
-            fail();
+            fail(ex.getMessage());
+        }
+        finally
+        {
+            final Boolean shutdownClean = client.shutdown().get(3, TimeUnit.SECONDS);
+            assertTrue(shutdownClean);
+        }
+    }
+
+    @Test
+    public void testThatConnectionsGetReturnedBeforeErrors() throws InterruptedException, ExecutionException, TimeoutException
+    {
+        assumeTrue(testLifecycle);
+
+        thrown.expect(ExecutionException.class);
+        thrown.expectMessage(containsString("no_type"));
+
+        RiakClient client = null;
+
+        try
+        {
+            client = new RiakClient(cluster);
+            cluster.start();
+
+            final Namespace namespace = new Namespace("doesnotexist", Integer.toString(random.nextInt()));
+            createAndStoreObject(client, new Location(namespace, "no_type"));
         }
         finally
         {

